src/main/scala/org/broadinstitute/hail/expr/FunctionRegistry.scala

```diff
@@ -0,0 +1,263 @@
+    lookup(name, TypeTag.funTypeFromArgs(typs)).map(_.retType)
+
+  def registerUnaryMethod[T, U](name: String, impl: T => U)
+    (implicit hrt: HailRep[T], hru: HailRep[U]) = {
+    bind(name, UnaryMethodType(hrt.typ), UnaryFun[T, U](hru.typ, impl))
+  }
+
+  def registerUnaryFun[T, U](name: String, impl: T => U)
+    (implicit hrt: HailRep[T], hru: HailRep[U]) = {
+    bind(name, UnaryFunType(hrt.typ), UnaryFun[T, U](hru.typ, impl))
+  }
+
+  def registerOptionUnaryMethod[T, U](name: String, impl: T => Option[U])
+    (implicit hrt: HailRep[T], hru: HailRep[U]) = {
+    bind(name, UnaryMethodType(hrt.typ), OptionUnaryFun[T, U](hru.typ, impl))
+  }
+
+  def registerOptionUnaryFun[T, U](name: String, impl: T => Option[U])
+    (implicit hrt: HailRep[T], hru: HailRep[U]) = {
+    bind(name, UnaryFunType(hrt.typ), OptionUnaryFun[T, U](hru.typ, impl))
+  }
+
+  def registerUnaryNAFilteredCollectionMethod[T, U](name: String, impl: TraversableOnce[T] => U)
+    (implicit hrt: HailRep[T], hru: HailRep[U]) = {
+    bind(name, UnaryMethodType(TArray(hrt.typ)), UnaryFun[IndexedSeq[_], U](hru.typ, { (ts: IndexedSeq[_]) =>
+      impl(ts.filter(t => t != null).map(_.asInstanceOf[T]))
+    }))
+    bind(name, UnaryMethodType(TSet(hrt.typ)), UnaryFun[Set[_], U](hru.typ, { (ts: Set[_]) =>
+      impl(ts.filter(t => t != null).map(_.asInstanceOf[T]))
+    }))
+  }
+
+  def registerBinary[T, U, V](name: String, impl: (T, U) => V)
+    (implicit hrt: HailRep[T], hru: HailRep[U], hrv: HailRep[V]) = {
+    bind(name, BinaryFunType(hrt.typ, hru.typ), BinaryFun[T, U, V](hrv.typ, impl))
+  }
+
+  def register3Ary[T, U, V, W](name: String, impl: (T, U, V) => W)
+    (implicit hrt: HailRep[T], hru: HailRep[U], hrv: HailRep[V], hrw: HailRep[W]) = {
+    bind(name, Arity3FunType(hrt.typ, hru.typ, hrv.typ), Arity3Fun[T, U, V, W](hrw.typ, impl))
+  }
+
+  def register4Ary[T, U, V, W, X](name: String, impl: (T, U, V, W) => X)
+    (implicit hrt: HailRep[T], hru: HailRep[U], hrv: HailRep[V], hrw: HailRep[W], hrx: HailRep[X]) = {
+    bind(name, Arity4FunType(hrt.typ, hru.typ, hrv.typ, hrw.typ), Arity4Fun[T, U, V, W, X](hrx.typ, impl))
+  }
+
+  registerOptionUnaryMethod("gt", { (x: Genotype) => x.gt })
+  registerOptionUnaryMethod("gtj", { (x: Genotype) => x.gt.map(gtx => Genotype.gtPair(gtx).j) })
+  registerOptionUnaryMethod("gtk", { (x: Genotype) => x.gt.map(gtx => Genotype.gtPair(gtx).k) })
+  registerOptionUnaryMethod("ad", { (x: Genotype) => x.ad.map(a => a: IndexedSeq[Int]) })
+  registerOptionUnaryMethod("dp", { (x: Genotype) => x.dp })
+  registerOptionUnaryMethod("od", { (x: Genotype) => x.od })
+  registerOptionUnaryMethod("gq", { (x: Genotype) => x.gq })
+  registerOptionUnaryMethod("pl", { (x: Genotype) => x.pl.map(a => a: IndexedSeq[Int]) })
+  registerOptionUnaryMethod("dosage", { (x: Genotype) => x.dosage.map(a => a: IndexedSeq[Double]) })
+  registerUnaryMethod("isHomRef", { (x: Genotype) => x.isHomRef })
```

Comment:
We should probably more clearly document the option vs non-option methods.  It maybe isn't 100% obvious `g.isHomRef` is false and not missing if `g` is no call.  The documentation only says, "true if this call is 0/0".

Doesn't need to be addressed here.  I created an issue: https://github.com/hail-is/hail/issues/829